File: 1.medium/array-912-solution.go
// https://leetcode.com/problems/sort-an-array/

package main

import "fmt"

func main() {
	seq := []int{9,8,7,6,5,4,3,2,1}
	fmt.Println(seq)

	res := sort(seq)
	fmt.Println(res)
}

func sort(seq []int) []int {
	mid := len(seq) / 2

	if mid == 0 {
		return seq
	}

	left := sort(seq[:mid])
	right := sort(seq[mid:])
	return merge(left, right)
}

func merge(left []int, right []int) []int {
	l := 0
	r := 0

	var res []int

	for l < len(left) && r < len(right) {
		if left[l] <= right[r] {
			res = append(res, left[l])
			l++
		} else {
			res = append(res, right[r])
			r++
		}
	}

	if l < len(left) {
		res = append(res, left[l:]...)
	}

	if r < len(right) {
		res = append(res, right[r:]...)
	}

	return res
}

File: LeetCode/1.medium/array-912-solution.js
// https://leetcode.com/problems/sort-an-array/

function merge(array) {
    const mid = Math.floor(array.length / 2)

    if (mid == 0) {
        return array
    }

    const left = merge(array.slice(0, mid))
    const right = merge(array.slice(mid))
    return sort(left, right)
}

function sort(left, right) {
    let l = 0
    let r = 0
    const result = []

    while (l < left.length && r < right.length) {
        if (left[l] <= right[r]) {
            result.push(left[l])
            l++
        } else {
            result.push(right[r])
            r++
        }
    }

    if (l < left.length) {
        result.push(...left.slice(l))
    }

    if (r < right.length) {
        result.push(...right.slice(r))
    }

    return result
}

const array = [9, 8, 7, 6, 5, 4, 3, 2, 1]
console.log(merge(array))
